src/appleseed/renderer/modeling/phasefunction/henyeyphasefunction.cpp

```diff
@@ -0,0 +1,323 @@
+        SamplingContext&            sampling_context,
+        const ShadingRay&           volume_ray,
+        const void*                 data,
+        float&                      distance
+        ) const APPLESEED_OVERRIDE
+    {
+        const InputValues* values = static_cast<const InputValues*>(data);
+
+        float ray_length = norm(volume_ray.m_dir) * (volume_ray.m_tmax - volume_ray.m_tmin);
+
+        // Sample distance.
+        sampling_context.split_in_place(1, 1);
+        const float s = sampling_context.next2<float>();
+        distance = foundation::sample_exponential_distribution_on_segment(
+            s, values->m_extinction_multiplier, 0.0f, ray_length);
+
+        // Return corresponding PDF value.
+        return foundation::exponential_distribution_on_segment_pdf(
+            distance, values->m_extinction_multiplier, 0.0f, ray_length);
+    }
+
+    virtual float sample(
+        SamplingContext&            sampling_context,
+        const ShadingRay&           volume_ray,
+        const void*                 data,
+        float                       distance,
+        foundation::Vector3f&       incoming) const APPLESEED_OVERRIDE
+    {
+        const InputValues* values = static_cast<const InputValues*>(data);
+        const foundation::Vector3f outgoing =
+            foundation::Vector3f(foundation::normalize(volume_ray.m_dir));
+        const foundation::Basis3f basis(outgoing);
+
+        //
+        // x = 1/2g * (1 + g^2 - [(1 - g^2) / (1 + g*s)]^2),
+        // where x is cos(phi) and s is uniform random sample from [-1; 1].
+        //
+
+        const float g = values->m_average_cosine;
+        const float sqr_g = g * g;
+
+        sampling_context.split_in_place(1, 2);
+        const float s = 2.0f * sampling_context.next2<float>() - 1.0f;
+
+        const float t = (1.0f - sqr_g) / (1.0f + g*s);
+        const float cosine = 0.5f * g * (1 + sqr_g - t * t);
+        const float sine = sqrt(1.0f - cosine * cosine);
+
+        const foundation::Vector2f tangent =
+            foundation::sample_circle_uniform(sampling_context.next2<float>());
+
+        incoming =
+            basis.get_tangent_u() * tangent.x * sine +
+            basis.get_tangent_v() * tangent.y * sine +
+            basis.get_normal()    * cosine;
+
+        // Evaluate PDF.
+
+        const float numerator = (1.0f - sqr_g);
+        const float denumerator = powf(1 + sqr_g - 2*g*cosine, -1.5f);
+
+        return foundation::RcpFourPi<float>() * numerator * denumerator;
+    }
+
+    virtual float evaluate(
+        const ShadingRay&           volume_ray,
+        const void*                 data,
+        float                       distance,
+        const foundation::Vector3f& incoming) const APPLESEED_OVERRIDE
+    {
+        const InputValues* values = static_cast<const InputValues*>(data);
+        const foundation::Vector3f outgoing =
+            foundation::Vector3f(foundation::normalize(volume_ray.m_dir));
```

Comment:
I think you could use the `foundation` namespace and get rid of all the qualifiers.